### ISSUE TITLE
Bugs in `.backward()` while training PCNet-C.

### ISSUE THREAD

**Issue by praeclarumjj3** [XiaohangZhan/deocclusion#38]
Hi! Thanks for sharing the excellent codebase. It's very helpful!

I came across an issue related to the **backward pass**  while training the PCNet-C network using the [PartialCompletionContentCGAN](https://github.com/XiaohangZhan/deocclusion/blob/9de459c15169929488ee0c68e5e03014a0e221a1/models/partial_completion_content_cgan.py#L145-L154) network. The lines responsible for the errors are:

```python
# update
self.optimD.zero_grad()
dis_loss.backward()
utils.average_gradients(self.netD)
self.optimD.step()


self.optim.zero_grad()
gen_loss.backward()
utils.average_gradients(self.model)
self.optim.step()
```

If we comment out either of the `.backward()` lines, the error goes away.

I am using Pytorch 1.8.1.

**Comment by praeclarumjj3** [comment]
The error stack is as follows:

```
main.py 48 <module>
main(args)

main.py 30 main
trainer.run()

trainer.py 125 run
self.train()

trainer.py 147 train
loss_dict = self.model.step()

partial_completion_content_cgan.py 152 step
gen_loss.backward()

tensor.py 245 backward
torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)

__init__.py 145 backward
Variable._execution_engine.run_backward(

RuntimeError:
one of the variables needed for gradient computation has been modified by an inplace operation: [torch.cuda.FloatTensor [1, 512, 4, 4]] is at version 2; expected version 1 instead. Hint: the backtrace further above shows the operation that failed to compute its gradient. The variable in question was changed in there or anywhere later. Good luck!
```

**Adding torch.autograd.set_detect_anomaly(True) gives some extra info:**

```
[W python_anomaly_mode.cpp:104] Warning: Error detected in DivBackward0. Traceback of forward call that caused the error:
  File "main.py", line 48, in <module>
    main(args)
  File "main.py", line 30, in main
    trainer.run()
  File "/workspace/jj/deocclusion/trainer.py", line 125, in run
    self.train()
  File "/workspace/jj/deocclusion/trainer.py", line 147, in train
    loss_dict = self.model.step()
  File "/workspace/jj/deocclusion/models/partial_completion_content_cgan.py", line 126, in step
    gen_fake, _ = self.netD(torch.cat([gen_input_fake, self.modal], dim=1))
  File "/workspace/jj/anaconda3/envs/pcnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/workspace/jj/deocclusion/utils/distributed_utils.py", line 18, in forward
    return self.module(*inputs, **kwargs)
  File "/workspace/jj/anaconda3/envs/pcnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/workspace/jj/deocclusion/models/backbone/discriminator.py", line 75, in forward
    conv5 = self.conv5(conv4)
  File "/workspace/jj/anaconda3/envs/pcnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/workspace/jj/anaconda3/envs/pcnet/lib/python3.8/site-packages/torch/nn/modules/container.py", line 119, in forward
    input = module(input)
  File "/workspace/jj/anaconda3/envs/pcnet/lib/python3.8/site-packages/torch/nn/modules/module.py", line 875, in _call_impl
    result = hook(self, input)
  File "/workspace/jj/anaconda3/envs/pcnet/lib/python3.8/site-packages/torch/nn/utils/spectral_norm.py", line 105, in __call__
    setattr(module, self.name, self.compute_weight(module, do_power_iteration=module.training))
  File "/workspace/jj/anaconda3/envs/pcnet/lib/python3.8/site-packages/torch/nn/utils/spectral_norm.py", line 92, in compute_weight
    weight = weight / sigma
 (function _print_stack)
```

**Comment by XiaohangZhan** [comment]
Hi,
This arises after pytorch1.5. See this solution:
https://stackoverflow.com/questions/62061703/runtimeerror-one-of-the-variables-needed-for-gradient-computation-has-been-modi

**Comment by praeclarumjj3** [comment]
> Hi,
> This arises after pytorch1.5. See this solution:
> https://stackoverflow.com/questions/62061703/runtimeerror-one-of-the-variables-needed-for-gradient-computation-has-been-modi

Hey, thanks for the reply! Yes, I saw that one. I have opened a [PR](https://github.com/XiaohangZhan/deocclusion/pull/39) fixing that part of the code. Please have a look at it.

**Comment by praeclarumjj3** [comment]
Pull Request #39 fixed this issue.
